### ISSUE TITLE
Performance issue with FFTs, numpy vs pytorch

### ISSUE THREAD

**Issue by chrismaliszewski** [pytorch/pytorch#44294]
I'm having a major problem with a performance of PyTorch while calculating FFT and things around it.

Code comparing times of both libraries:
```python
import torch
import numpy as np
import time

class Timer:
    def __init__(self, precision=7):
        self._start_time = {}
        self.precision = precision
        self._text = None

    def start(self, text=''):
        """Start a new timer"""
        self._start_time[str(text)] = time.perf_counter()

    def stop(self, text='', printable=True):
        """Stop the timer, and report the elapsed time"""
        if str(text) not in self._start_time:
            print(f"Timer is not running. Use .start() to start it")

        elapsed_time = time.perf_counter() - self._start_time[str(text)]
        self._start_time[str(text)] = None
        if printable: print(str(text)+' elapsed time: '+str(np.round(elapsed_time, self.precision))+' seconds')
        
        return elapsed_time
    
def calc_numpy(data, shape_fft):
        t = Timer()
        """ init required data """
        t.start('zero')
        dataFFT = np.zeros(shape_fft)
        t.stop('zero')
        
        """ rot data """
        t.start('rot')
        data = np.rot90(data)
        t.stop('rot')
        
        t.start('loop')
        for ii, num in enumerate(np.arange(0,shape_fft[0],1)):
            """ take a diagonal """
            # t.start('diag')
            diag = np.diag(data,ii)
            # t.stop('diag')
            
            """ FFT the diagonal spectrum """
            # t.start('fft')
            fftdiag = np.abs(np.fft.fft(diag, n=shape_fft[1] * 2))
            # t.stop('fft')
            
            # t.start('cut')
            dataFFT[ii, :]  = fftdiag[0:shape_fft[1]]
            # t.stop('cut')
        t.stop('loop')
        
        return dataFFT

def calc_pytorch(data, shape_fft):
        t = Timer()
        """ init required data """
        """ NOTE: pytorch fft does not allow to set a shape of an output,
        therefore zero-padding must be done manually,
        diag_zero is used to speed-up the process of zero-padding by initializing 
        the required shape tensor instead of grabbing data and zero-padding it on the way """
        t.start('zero')
        dataFFT = torch.zeros(shape_fft)
        shape_data = data.shape[0]
        diag_zero = torch.zeros((shape_data * 2 - 1))
        t.stop('zero')
        
        """ rot data """
        t.start('rot')
        data = torch.rot90(data)
        t.stop('rot')
        
        t.start('loop')
        for ii, num in enumerate(torch.arange(0,shape_fft[0],1)):
            """ take a diagonal spectrum """
            """ NOTE, diag_zero is used to speed-up the process of zero-padding"""
            # t.start('diag_p')
            diag = diag_zero
            diag[:shape_data - ii] = torch.diag(data,ii)
            # t.stop('diag_p')
            
            """ FFT the diagonal spectrum """
            """ torch fft returns data in a way it is not possible to easily calc abs """
            """ it is supposed to be fixed in the future version but does not work as for now """
            # t.start('fft_p')
            fftdiag = torch.rfft(diag, 1)
            dataFFT[ii, :] = torch.sqrt(torch.pow(fftdiag[:, 0], 2) + torch.pow(fftdiag[:, 1], 2))
            # t.stop('fft_p')
        t.stop('loop')
        
        return dataFFT

shape_fft = (50, 256)
shape_data = (256, 256)
data = np.random.rand(shape_data[0], shape_data[1])

t = Timer()
t.start('NUMPY TIME')
calc_numpy(data, shape_fft)
t.stop('NUMPY TIME')

data = torch.Tensor(data)
t.start('PYTORCH TIME')
calc_pytorch(data, shape_fft)
t.stop('PYTORCH TIME')
```

Results:
```
zero elapsed time: 1.32e-05 seconds
rot elapsed time: 4.75e-05 seconds
loop elapsed time: 0.0012882 seconds
NUMPY TIME elapsed time: 0.0022629 seconds
zero elapsed time: 3.97e-05 seconds
rot elapsed time: 0.0004176 seconds
loop elapsed time: 0.0057724 seconds
PYTORCH TIME elapsed time: 0.0070718 seconds
```

In general, PyTorch is 3-4x slower than NumPy. 

The main problems lay in the following things:
1. FFT which does not allow to set output shape param; because of that, the data must be prepared accordingly by zero-padding beforehand which takes time to initialize required data structures and set values. NOTE: The way I created the zero-padding part is faster than using PyTorch zero-padding functions. 
2. Providing the output of FFT to abs function returns wrong values. I read that it will be fixed in the future release. I must calculate it myself at this moment.

The results of both ways differ a bit but I don't mind.

How can I speed up the process? Any ideas? 

cc @mruberry @rgommers @VitalyFedyunin @ngimel

**Comment by SsnL** [comment]
Use `.norm()` to calculate abs.
```
zero elapsed time: 8.12e-05 seconds
rot elapsed time: 7.61e-05 seconds
loop elapsed time: 0.1728914 seconds
NUMPY TIME elapsed time: 0.1735949 seconds
zero elapsed time: 0.0002396 seconds
rot elapsed time: 0.0005161 seconds
loop elapsed time: 0.0633892 seconds
PYTORCH TIME elapsed time: 0.0647402 seconds
```

**Comment by chrismaliszewski** [comment]
Great. Thanks for that @SnnL! Additionally, the results are not the same. One step at a time. 

Unfortunately, timing is still the same, bad. 
```
zero elapsed time: 1.28e-05 seconds
rot elapsed time: 5.27e-05 seconds
loop elapsed time: 0.0015067 seconds
NUMPY TIME elapsed time: 0.0018481 seconds
-----------------
zero elapsed time: 3.07e-05 seconds
rot elapsed time: 0.0003838 seconds
loop elapsed time: 0.0080958 seconds
PYTORCH TIME elapsed time: 0.0093602 seconds
```
How did you achieve your great timing? 

I don't have GPU to test so I moved the code to Colab (https://colab.research.google.com/drive/1XKAX9Gs8CGWxpDImFRPxTlrxkUsXzpzQ?usp=sharing) and it returns similar times:
```
zero elapsed time: 2.01e-05 seconds
rot elapsed time: 6.16e-05 seconds
loop elapsed time: 0.0018384 seconds
NUMPY TIME elapsed time: 0.0022646 seconds
-----------------
DEVICE:  cuda
zero elapsed time: 6.1e-05 seconds
rot elapsed time: 0.0001687 seconds
loop elapsed time: 0.009126 seconds
PYTORCH TIME elapsed time: 0.0098478 seconds
```

**Comment by izdeby** [comment]
Please, use our [forums](https://discuss.pytorch.org/) for discussion.

**Comment by rgommers** [comment]
@chrismaliszewski your code isn't the same for numpy as for pytorch, so it's hard to compare. For example, you're using `torch.rfft` and comparing vs. `np.fft.fft` - both numpy and pytorch have `fft` and `rfft`. I suggest rewriting your code to benchmark as a single function:

```
def calc(lib, data, shape_fft):
    """`lib` is either `torch` or `np`"""
    ....
    data = lib.rot90(data)
    fftdiag = lib.rfft(...)
```

This should be possible with gh-43011, which adds numpy-compatible 1-D transforms. (if you can't build from source, watch that PR and install from a nightly when it lands - should be soon)

There's a few notes on multi-dimensional FFT performance in https://github.com/pytorch/pytorch/issues/42175#issuecomment-673580206 which may be relevant as well. Cc @peterbell10.

**Comment by peterbell10** [comment]
Timing with gh-43011 still shows a performance drop with pytorch:
```
zero elapsed time: 3.07e-05 seconds
rot elapsed time: 5.59e-05 seconds
loop elapsed time: 0.001341 seconds
NUMPY TIME elapsed time: 0.0015816 seconds
zero elapsed time: 8.81e-05 seconds
rot elapsed time: 0.0002193 seconds
loop elapsed time: 0.0037979 seconds
PYTORCH TIME elapsed time: 0.00421 seconds
```

However, I don't think this is a very good benchmark. It does many small non-vectorized 1D transforms which is not what tensor libraries are best at. It's also conflating `diag`, `fft`, `abs` and copy-assign into one single benchmark.

If I use `timeit` to isolate each line:
* diag is ~3 us for numpy and ~5-6 us for pytorch
* fft is ~10 us for numpy and 20 us for pytorch
* abs is ~9 us for numpy and 12 us for pytorch
* assignment is ~1us in numpy and 11 us for pytorch.

Every operator has at least 3 us overhead with pytorch. That's pretty bad, but for large tensors I'm sure it balances out a bit better.  FFT is about 2x slower for this small 1D tensor, but for a `256 X 512` FFT pytorch is 2x faster than NumPy, even single threaded.

The biggest standout is copying from one slice to another which is an order of magnitude slower in pytorch.

**Comment by chrismaliszewski** [comment]
Thank you, @rgommers, and especially you, @peterbell10, for deep analysis. You confirmed my worries that not much can be done. I'll use NumPy for calculating data and, then, convert the results to Tensor. 

I think it summarizes the issue. 
Thanks, again!
